cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java

```diff
@@ -626,37 +628,43 @@ private void send(Event event) {
     // "see" non-stargate node events
 
     @Override
-    public void onJoinCluster(InetAddressAndPort endpoint) {
-      if (server.persistence.isRpcReady(endpoint)) endpointsPendingJoinedNotification.add(endpoint);
-      else
-        onTopologyChange(
-            endpoint, Event.TopologyChange.newNode(server.persistence.getNativeAddress(endpoint)));
+    public void onJoinCluster(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      if (server.isRpcReady.get()) endpointsPendingJoinedNotification.add(endpointWithPort);
+      else onTopologyChange(endpointWithPort, Event.TopologyChange.newNode(endpointWithPort));
     }
 
     @Override
-    public void onLeaveCluster(InetAddressAndPort endpoint) {
-      onTopologyChange(
-          endpoint,
-          Event.TopologyChange.removedNode(server.persistence.getNativeAddress(endpoint)));
+    public void onLeaveCluster(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      onTopologyChange(endpointWithPort, Event.TopologyChange.removedNode(endpointWithPort));
     }
 
     @Override
-    public void onMove(InetAddressAndPort endpoint) {
-      onTopologyChange(
-          endpoint, Event.TopologyChange.movedNode(server.persistence.getNativeAddress(endpoint)));
+    public void onMove(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      onTopologyChange(endpointWithPort, Event.TopologyChange.movedNode(endpointWithPort));
     }
 
     @Override
-    public void onUp(InetAddressAndPort endpoint) {
-      if (endpointsPendingJoinedNotification.remove(endpoint)) onJoinCluster(endpoint);
+    public void onUp(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      if (endpointsPendingJoinedNotification.remove(endpointWithPort)) onJoinCluster(endpoint);
+
+      onStatusChange(endpointWithPort, Event.StatusChange.nodeUp(endpointWithPort));
+    }
 
-      onStatusChange(
-          endpoint, Event.StatusChange.nodeUp(server.persistence.getNativeAddress(endpoint)));
+    public void onDown(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      onStatusChange(endpointWithPort, Event.StatusChange.nodeDown(endpointWithPort));
     }
 
-    public void onDown(InetAddressAndPort endpoint) {
-      onStatusChange(
-          endpoint, Event.StatusChange.nodeDown(server.persistence.getNativeAddress(endpoint)));
+    private InetAddressAndPort withPort(InetAddress endpoint) {
+      return InetAddressAndPort.getByAddressOverrideDefaults(
+          endpoint,
+          server.useSSL
+              ? TransportDescriptor.getNativeTransportPortSSL()
+              : TransportDescriptor.getNativeTransportPort());
```

Comment:
Those are the ports of the current instance though. Are we sure that the remote node will use the same ones?